### PR TITLE
fix(xjy): add decode text function and  decode_loss_mode option of reconstruction loss for jericho

### PR DESCRIPTION
**Pull request by xiongjyu**

### CODE REVIEW

File: lzero/model/unizero_world_models/tokenizer.py [hunk 1]
```diff
@@ -53,19 +59,34 @@ def __init__(self, encoder=None, decoder_network=None, with_lpips: bool = False,
...
-        vocab_size = self.decoder_network.embed_tokens.weight.size(0)
-        self.decoder_network.lm_head = nn.Linear(
-            self.decoder_network.config.d_model, vocab_size, bias=False
-        )
-        self.decoder_network.lm_head.weight = self.decoder_network.embed_tokens.weight
+        # vocab_size = self.decoder_network.embed_tokens.weight.size(0)
+        # self.decoder_network.lm_head = nn.Linear(
+        #     self.decoder_network.config.d_model, vocab_size, bias=False
+        # )
+        # self.decoder_network.lm_head.weight = self.decoder_network.embed_tokens.weight
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:69]: 这里不用的代码直接删除了吧

File: lzero/model/unizero_world_models/tokenizer.py [hunk 2]
```diff
@@ -112,91 +133,63 @@ def decode_to_obs(self, embeddings: torch.Tensor) -> torch.Tensor:
...
-        enc_attn_mask = torch.ones(embeddings.size(0),1,
-                                   dtype=torch.long, device=embeddings.device)
-
-        outputs = self.decoder_network(
-            input_ids = decoder_input_ids,
-            attention_mask = dec_attn_mask,
-            encoder_hidden_states = embeddings,
-            encoder_attention_mask = enc_attn_mask,
+
+        embeddings = self.projection_layer(embeddings)     # (B',1,E)
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:163]: 为什么是B' 更新一下注释

File: lzero/model/unizero_world_models/tokenizer.py [hunk 3]
```diff
@@ -112,91 +133,63 @@ def decode_to_obs(self, embeddings: torch.Tensor) -> torch.Tensor:
...
-        shifted[:, 0] = start_token_id
-        shifted.masked_fill_(shifted == -100, pad_token_id)
-        return shifted
-
-    def decode_to_language_logits(
-        self, embeddings: torch.Tensor, target_ids: torch.Tensor,
-        pad_token_id: int = 0, decoder_start_token_id: int = 0) -> torch.Tensor:
-
+    def decode_to_reconstruction_outputs(self, embeddings: torch.Tensor, target_ids: torch.Tensor) -> torch.Tensor:
+        """
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:137]: 这个方法整体的逻辑在这里增加一下英文版的介绍吧

File: lzero/model/unizero_world_models/tokenizer.py [hunk 4]
```diff
@@ -112,91 +133,63 @@ def decode_to_obs(self, embeddings: torch.Tensor) -> torch.Tensor:
...
-            decoder_start_token_id (int): 解码时的起始 token id。
-            eos_token_id (int): 序列结束 token 的 id。
-            sampling (bool): 是否采用采样解码；默认 False，采用贪心解码。
-            top_k (Optional[int]): 使用 top-k 采样时的 k 值，如果提供则会对 logits 进行 top-k 过滤。
-            top_p (Optional[float]): 使用 nucleus（top-p）采样时的 p 值，如果提供则会对 logits 进行 top-p 过滤。
-        
-        返回:
-            List[List[int]]: 生成的 token 序列，每个子列表代表一个 batch 内序列 (剔除起始 token)。
+        Overview:
+            Decode embeddings into plain text using decoder's generate method.
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:185]: 这里也是增加一下整体介绍

File: lzero/model/unizero_world_models/tokenizer.py [hunk 5]
```diff
@@ -217,141 +210,29 @@ def decode_to_language_logits_for_inference(
...
-                
-                # 更新 is_finished 标志：若当前 token 为 eos_token_id，则标记该序列结束
-                is_finished |= next_token.squeeze(-1).eq(eos_token_id)
-                
-                # 如果所有序列均已生成 EOS，则提前退出循环
-                if is_finished.all():
-                    break
-
+            generated_t5_ids = self.decoder_network.generate(
+                encoder_outputs=encoder_outputs_tuple,
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:225]: 增加一下embeddings作为输入，其内部的简单工作原理

File: lzero/model/unizero_world_models/tokenizer.py [hunk 6]
```diff
@@ -217,141 +210,29 @@ def decode_to_language_logits_for_inference(
...
+                attention_mask=encoder_attention_mask,
+                max_length=max_length
+            )
+            generated_text = self.decoder_network_tokenizer.batch_decode(
+                generated_t5_ids, skip_special_tokens=True)
+            
             # 返回结果时去掉起始的 decoder_start_token_id，并将结果转换为 CPU 上的 list 类型
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:232]: 中文注释改成英文版本

File: lzero/model/unizero_world_models/tokenizer.py [hunk 7]
```diff
@@ -112,91 +133,63 @@ def decode_to_obs(self, embeddings: torch.Tensor) -> torch.Tensor:
...
 
-        embeddings = self.projection_layer(embeddings)     # (B',1,E)
+        text_list = self.decode_to_plain_text(target_ids)
+        t5_target_ids = self.decoder_network_tokenizer(text_list, 
+                                                       padding="max_length",
+                                                       truncation=True, 
+                                                       max_length=512, 
+                                                       return_tensors="pt")
+        labels = t5_target_ids.input_ids
+        labels[labels == self.decoder_network_tokenizer.pad_token_id] = -100
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:160]: 增加这里不用target_ids 而用转换后的id的原因

File: lzero/model/unizero_world_models/world_model.py [hunk 8]
```diff
@@ -1357,27 +1358,23 @@ def compute_loss(self, batch, target_tokenizer: Tokenizer = None, inverse_scalar
...
-            )
-
-            latent_recon_loss = self.tokenizer.lm_reconstruction_loss(
-                batch['observations'],
-                reconstructed_logits,
-                ignore_index = pad_id,     # 仍然忽略 <pad>
-            )
-
+            # 重建损失： 预测下一latent的重建损失（经过backbone）
+            # input -> encoder -> backbone(unizero) -> decoder -> latent_recon_loss
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/world_model.py:1362]: 通过一个option控制decode loss的位置

File: lzero/model/unizero_world_models/world_model.py [hunk 9]
```diff
@@ -1420,20 +1408,6 @@ def compute_loss(self, batch, target_tokenizer: Tokenizer = None, inverse_scalar
...
-        # predict_rewards = inverse_scalar_transform_handle(outputs.logits_rewards.reshape(-1, 101)).reshape(batch['observations'].shape[0], batch['observations'].shape[1], 1)
-        # import pdb; pdb.set_trace()
-        # visualize_reward_value_img_policy(original_images, reconstructed_images, target_predict_value, true_rewards, target_policy, predict_value, predict_rewards, predict_policy, not_plot_timesteps=[], suffix='pong_H10_H4_0613')
-
-        # visualize_reward_value_img_policy(original_images, reconstructed_images, target_predict_value, true_rewards, target_policy, predict_value, predict_rewards, predict_policy, not_plot_timesteps=list(np.arange(4,60)), suffix='visual_match_memlen1-60-15/one_success_episode')
-        # visualize_reward_value_img_policy(original_images, reconstructed_images, target_predict_value, true_rewards, target_policy, predict_value, predict_rewards, predict_policy, not_plot_timesteps=list(np.arange(4,60)), suffix='visual_match_memlen1-60-15/one_fail_episode')
-        #  ========== for visualization ==========
-
-        # For training stability, use target_tokenizer to compute the true next latent state representations
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/world_model.py:1436]: 这一句注释保留

File: lzero/worker/muzero_collector.py [hunk 10]
```diff
@@ -546,6 +542,12 @@ def collect(self,
                     groundtrut_next_text[env_id] = self._env._envs[env_id].tokenizer.decode(valid_input_ids, skip_special_tokens=True)
                         
                     text_bleu = compute_bleu(reference=groundtrut_next_text[env_id], prediction=pred_next_text[env_id])
+                    
+                    # 是否输出高bleu的文本比较结果，用来检验decode下一latent的效果
```
Comment (puyuan1996) [inline lzero/worker/muzero_collector.py:546]: 通过option控制 默认输出

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: lzero/worker/muzero_collector.py (hunk 10 above)
Comment (xiongjyu) [reply lzero/worker/muzero_collector.py:546]: 老师，这个我觉得就不加个option了，加了感觉参数变多了会有点复杂，所以我改成了默认让他输出，而且这部分只是print应该不会影响速度

File: lzero/mcts/buffer/game_buffer.py [hunk 11]
```diff
@@ -152,13 +152,14 @@ def _sample_orig_data(self, batch_size: int) -> Tuple:
...
             # [0, game_segment_length - num_unroll_steps] to avoid padded data.
 
-            # TODO: Consider increasing `self._cfg.game_segment_length` to ensure sampling efficiency.
-            # if pos_in_game_segment >= self._cfg.game_segment_length - self._cfg.num_unroll_steps:
-            #     pos_in_game_segment = np.random.choice(self._cfg.game_segment_length - self._cfg.num_unroll_steps, 1).item()
-
-            # NOTE: Sample the init position from the whole segment, but not from the padded part
-            if pos_in_game_segment >= self._cfg.game_segment_length:
-                pos_in_game_segment = np.random.choice(self._cfg.game_segment_length, 1).item()
+            if self._cfg.action_type == 'varied_action_space':
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer.py:155]: 可以在PR description中增加简要的描述和实验结果，方便后续继续开发哈
